### PR TITLE
Support multiple entry points

### PR DESCRIPTION
**Pull request by tungld**
This patch adds support for multiple entry points by adding the following C functions:
-  `const char *omInputSignature(entry_point)` and `const char *omOutputSignature(entry_point)` to query the input/output signature of the given entry point, respectively.
- `const char **omQueryEntryPoints()` to query all entry point names in a model. It returns an array of strings, e.g. `["run_add", "run_sub", NULL]`. NULL is added as a terminator.

Example with C++ ExecutionSession:
```C
// Model has a single entry point. Entry point defaults to `run_main_graph`.
ExecutionSession sess = ExecutionSession(sharedLib);
std::string inSig = sess.inputSignature();
sess.run(...)

// Model has multiple entry points.
ExecutionSession sess = ExecutionSession(sharedLib, /*defaultEntryPoint=*/false);
std::string* entryPoints = sess.queryEntryPoints();
//
sess.setEntryPoint(entryPoints[0]);
std::string inSig = sess.inputSignature(); // query the input signature of the current entry point.
sess.run(...)
//
sess.setEntryPoint(entryPoints[1]);
std::string inSig = sess.inputSignature(); // query the input signature of the current entry point.
sess.run(...)
```

Python ExecutionSession is similar to C++ ExecutionSession.

This patch does not add a JNI wrapper for `omQueryEntryPoints`.

### CODE REVIEW

**Comment by gongsu832** [comment]
@tungld With multiple entries, should we have an API call to query the list of entry names? e.g., something like
```
const char **omQueryEntry();
```
which returns an array of strings terminated by NULL, e.g., something like
```
{ "run_main_graph", "test_add", "test_relu", NULL };
```

**Comment by tungld** [comment]
> @tungld With multiple entries, should we have an API call to query the list of entry names?

Good idea. I will implement that API. Just thinking whether we should return an array of strings or a JSON string. JSON is using for signatures, so it looks more consistent.

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 1]
```diff
@@ -1895,6 +1849,202 @@ void mlir::checkConstantOutputs(
...
+  // When there is only a single entry point function in a model, use
+  // DEFAULT_DYN_ENTRY_POINT.
+  if (entryPointNames.size() == 1) {
+    entryPointNames[0] = DEFAULT_DYN_ENTRY_POINT;
+    entryPointNames[0].push_back('\0'); // null terminate the string.
+  }
+}
+
+void mlir::genSignatureFunction(ModuleOp module,
+    ArrayRef<std::string> entryPointNames, ArrayRef<std::string> inSignatures,
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:1889]: `entryPointNames`, `inSignatures ` and `outSignatures ` should not be modified by this function, suggest to make them const.

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 2]
```diff
@@ -1895,6 +1849,202 @@ void mlir::checkConstantOutputs(
...
+    mlir::StringAttr inSigAttr =
+        mlir::StringAttr::get(context, inSignatures[i]);
+    Type inSigArrayType =
+        LLVM::LLVMArrayType::get(i8Type, inSignatures[i].size());
+    std::string inSigVarName = entryVarName + "_in_sig";
+    LLVM::GlobalOp inSigOp = b.create<LLVM::GlobalOp>(loc, inSigArrayType,
+        /*isConstant=*/true, LLVM::Linkage::External, inSigVarName, inSigAttr);
+    inSigOps.emplace_back(inSigOp);
+
+    // Global constants for input signatures.
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:1937]: "input signatures" -> "output signatures"

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 3]
```diff
@@ -1895,6 +1849,202 @@ void mlir::checkConstantOutputs(
...
+    mlir::StringAttr inSigAttr =
+        mlir::StringAttr::get(context, inSignatures[i]);
+    Type inSigArrayType =
+        LLVM::LLVMArrayType::get(i8Type, inSignatures[i].size());
+    std::string inSigVarName = entryVarName + "_in_sig";
+    LLVM::GlobalOp inSigOp = b.create<LLVM::GlobalOp>(loc, inSigArrayType,
+        /*isConstant=*/true, LLVM::Linkage::External, inSigVarName, inSigAttr);
+    inSigOps.emplace_back(inSigOp);
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:1935]: This code section looks similar to the one above (lines 1918-1925) and below (lines 1938-1946). Suggest to create a lambda or a templated function to avoid code duplication.

File: src/Runtime/ExecutionSession.cpp [hunk 4]
```diff
@@ -96,10 +92,12 @@ OMTensorList *ExecutionSession::run(OMTensorList *input) {
   return _entryPointFunc(input);
 }
 
-std::string ExecutionSession::inputSignature() { return _inputSignatureFunc(); }
+std::string ExecutionSession::inputSignature() {
```
Comment (etiotto) [inline src/Runtime/ExecutionSession.cpp:95]: ExecutionSession::inputSignature() --> ExecutionSession::inputSignature() const. Similar comment for `ExecutionSession::outputSignature()`

File: src/Runtime/ExecutionSession.hpp [hunk 5]
```diff
@@ -25,7 +25,7 @@
 namespace onnx_mlir {
 
 typedef OMTensorList *(*entryPointFuncType)(OMTensorList *);
-typedef const char *(*signatureFuncType)();
+typedef const char *(*signatureFuncType)(const char *);
```
Comment (etiotto) [inline src/Runtime/ExecutionSession.hpp:28]: Suggest to use a using declaration rather than typedef, given this is C++ code.

File: src/Runtime/PyExecutionSession.cpp [hunk 6]
```diff
@@ -162,6 +162,24 @@ std::vector<py::array> PyExecutionSession::pyRun(
   return outputPyArrays;
 }
 
+void PyExecutionSession::pySetEntryPoint(std::string entryPointName) {
+  setEntryPoint(entryPointName);
+  return;
+}
+
+std::vector<std::string> PyExecutionSession::pyQueryEntryPoint() {
```
Comment (gongsu832) [inline src/Runtime/PyExecutionSession.cpp:170]: `ExecutionSession.{c|h}pp` use `queryEntryPoints()` so this should be `pyQueryEntryPoints()` to be consistent.

Comment (tungld) [reply src/Runtime/PyExecutionSession.cpp:170]: Yes. Thanks!

**Comment by tungld** [comment]
Addressed all comments from Gong and Ettore. Thanks!

**Comment by gongsu832** [comment]
> @gongsu832 I do not write a JNI wrapper for omQueryEntryPoints. I think you are the best person to do that. Thanks!

@tungld Yes I will add the call for Java API. Thanks.

File: src/Runtime/PyExecutionSession.cpp [hunk 7]
```diff
@@ -162,6 +162,24 @@ std::vector<py::array> PyExecutionSession::pyRun(
   return outputPyArrays;
 }
 
+void PyExecutionSession::pySetEntryPoint(std::string entryPointName) {
+  setEntryPoint(entryPointName);
+  return;
```
Comment (etiotto) [inline src/Runtime/PyExecutionSession.cpp:167]: no need to return here

File: include/onnx-mlir/Runtime/OMEntryPoint.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+extern "C" {
+#endif
+
+/**
+ * \brief Return all entry point names in a model.
+ *
+ * @return an array of strings. The array ends with NULL. For example:
+ * ["run_add", "run_sub", NULL].
+ */
+const char **omQueryEntryPoints();
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply include/onnx-mlir/Runtime/OMEntryPoint.h:27]: NULL terminated string array is fairly common. Unless there are many use cases where you only want the number of entry points but not the entry points themselves, this would be the simplest API.

[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply include/onnx-mlir/Runtime/OMEntryPoint.h:27]: It is fine, but the type may need to be more `const char * const *` as you have pointer to a constant array of pointers, each of which points to a constant char string.

[the remark this answers is not in the document]
Comment (tungld) [reply include/onnx-mlir/Runtime/OMEntryPoint.h]: Done. I changed to `const char * const *`. Keep using NULL-terminated arrays. Thanks!

File: src/Runtime/PyExecutionSession.cpp (hunk 7 above)
Comment (tungld) [reply src/Runtime/PyExecutionSession.cpp]: Done. Thanks!

**Comment by AlexandreEichenberger** [comment]
I added some high level comments to the related issue #1183

File: include/onnx-mlir/Runtime/OMSignature.h [hunk 9]
```diff
@@ -19,18 +19,22 @@ extern "C" {
...
 /**
- * \brief Return the model's input signature as a JSON string.
+ * \brief Return the model's input signature as a JSON string for the given
+ * entry point name.
  *
+ * @param entry point name
  * @return pointer to input signature JSON string
  */
-const char *omInputSignature();
+const char *omInputSignature(const char *);
```
[the remark this answers is not in the document]
Comment (tungld) [reply include/onnx-mlir/Runtime/OMSignature.h]: Done for both omInputSignature and omOutputSignature. Thanks!

File: src/Runtime/ExecutionSession.hpp [hunk 10]
```diff
@@ -41,10 +41,17 @@ class ExecutionSession {
   // tensor lists.
   OMTensorList *run(OMTensorList *input);
 
+  // Set entry point for this session.
+  void setEntryPoint(std::string);
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Runtime/ExecutionSession.hpp]: Fixed this.

**Comment by AlexandreEichenberger** [comment]
Copied here the question I posted on the related issue.

@tungld,

I think we must be very clear on what the "entry point" are. It may be self evident to us now, but this is important for the user as they are exposed to these names. Is it correct that these "entry point" are actually the symbol of the compute function? If so, they are exposed to the user as if you have a statically linked compute function, like in many of our examples, we expect the user to define and then call entry_point(inputTensorList), and if the user is working with dynamic libraries, the user is expected to do a `dlsym(entry_point).

I agree with @gongsu832 that we should have functions to get the signatures that passes the entry point name. But this starts to feel like we have a data structure for each entry point.

struct EntryPointStruct {
  char *entryPointName;
  char *inputSignature;
  char *outputSignature;
}
That feel preferable to having hardwired functions to get the signatures, as if we have multiple entry point, we must start to enumerate/generate to get the right function for a solution 1

if (entry_point_name == "test_add")
    run_test_add_in_signature();
else if (entry_point_name == "test_relu")
    run_test_relu_in_signature();
(since the name are constants, probably test the address of the constant is enough).

**Comment by gongsu832** [comment]
@AlexandreEichenberger The main reason the signature APIs are done in the current way as oppose to using a `struct EntryPointStruct` is to maintain backward compatibility.

**Comment by AlexandreEichenberger** [comment]
@gongsu832 we have no signature in our current design, and it feel that a struct would be a very efficient way to implement one

**Comment by gongsu832** [comment]
We do. WMLz folks have been using the Java signature APIs, which is a direct JNI call to the C signature APIs, for quite a while.

File: src/Runtime/ExecutionSession.cpp [hunk 11]
```diff
@@ -96,10 +97,27 @@ OMTensorList *ExecutionSession::run(OMTensorList *input) {
   return _entryPointFunc(input);
 }
 
-std::string ExecutionSession::inputSignature() { return _inputSignatureFunc(); }
+void ExecutionSession::setEntryPoint(const std::string &entryPointName) {
```
Comment (AlexandreEichenberger) [inline src/Runtime/ExecutionSession.cpp:100]: If a user does not want the default entry point, and does not call `setEntryPoint`, then the `_entryPointFunc` remains null, and is unconditionally called in the `run` function. I suggest to add something similar to the code below inside the `run` function.
```C
if (!_entryPointFunc) {
     std::stringstream errStr;
     errStr << "Must set the entry point before calling run function"<< std::endl;
     throw std::runtime_error(errStr.str());
   }
```
Would do the same for the other signature functions too.

File: src/Runtime/ExecutionSession.hpp [hunk 12]
```diff
@@ -41,10 +41,17 @@ class ExecutionSession {
   // tensor lists.
   OMTensorList *run(OMTensorList *input);
 
+  // Set entry point for this session.
```
Comment (AlexandreEichenberger) [inline src/Runtime/ExecutionSession.hpp:44]: I would add that this call is required before calling run/input/output signatures when default entry point is not selected. To that effect, I would move this def above the definition of run, so that the sequence is clearer.

File: src/Runtime/ExecutionSession.cpp [hunk 13]
```diff
@@ -96,10 +97,27 @@ OMTensorList *ExecutionSession::run(OMTensorList *input) {
...
+  }
+  _entryPointName = entryPointName;
+}
+
+const std::string *ExecutionSession::queryEntryPoints() const {
+  return (const std::string *)_queryEntryPointsFunc();
+}
+
+const std::string ExecutionSession::inputSignature() const {
+  return _inputSignatureFunc(_entryPointName.c_str());
```
Comment (AlexandreEichenberger) [inline src/Runtime/ExecutionSession.cpp:116]: if the entry point was not set, then what happens in the `_inputSignatureFunc`? Should we test this here?

File: include/onnx-mlir/Runtime/OMSignature.h [hunk 14]
```diff
@@ -19,18 +19,20 @@ extern "C" {
...
 
 /**
- * \brief Return the model's input signature as a JSON string.
+ * \brief Return the input signature of the given entry point as a JSON string.
  *
+ * @param entry point name
  * @return pointer to input signature JSON string
  */
-const char *omInputSignature();
+const char *omInputSignature(const char *entryPointName);
```
Comment (AlexandreEichenberger) [inline include/onnx-mlir/Runtime/OMSignature.h:27]: Can you briefly explain how the entryPointName is used to locate the right signature? I tried to read the LLVM changes but I could not really get the big picture.

File: test/backend-cpp/ModelBuilder.cpp [hunk 15]
```diff
@@ -63,8 +63,8 @@ bool ModelBuilder::runAndVerifyTest(std::vector<OMTensorUniquePtr> &inputs,
   assert(!inputs.empty() && "Expecting valid inputs");
 
   // Run the test code.
-  onnx_mlir::ExecutionSession execSession(
-      getSharedLibName(sharedLibBaseName), "run_main_graph");
+  onnx_mlir::ExecutionSession execSession(getSharedLibName(sharedLibBaseName));
+  execSession.setEntryPoint("run_main_graph");
```
Comment (AlexandreEichenberger) [inline test/backend-cpp/ModelBuilder.cpp:67]: Can you confirm that this call is redundant as the exec session uses by default the "run_main_graph" name.

File: test/numerical/numerical.def [hunk 16]
```diff
@@ -1,5 +1,5 @@
-EXPORTS 
-run_main_graph
-omInputSignature
-omOutputSignature
-
+EXPORTS
```
Comment (AlexandreEichenberger) [inline test/numerical/numerical.def:1]: Can you double check that these changes are made to all of the ".def" files? Thanks

File: src/Runtime/ExecutionSession.hpp [hunk 17]
```diff
@@ -24,15 +24,15 @@
...
+using entryPointFuncType = OMTensorList *(*)(OMTensorList *);
+using queryEntryPointsFuncType = const char **(*)();
+using signatureFuncType = const char *(*)(const char *);
 using OMTensorUniquePtr = std::unique_ptr<OMTensor, decltype(&omTensorDestroy)>;
 
 class ExecutionSession {
 public:
   ExecutionSession(std::string sharedLibPath);
-  ExecutionSession(std::string sharedLibPath, std::string entryPointName);
+  ExecutionSession(std::string sharedLibPath, bool defaultEntryPoint);
```
Comment (AlexandreEichenberger) [inline src/Runtime/ExecutionSession.hpp:35]: nit, why not just one `ExecutionSession(std::string sharedLibPath, bool defaultEntryPoint=true);`

**Comment by AlexandreEichenberger** [comment]
My point above was on whether the get signatures were efficiently implemented or not (and maybe having a struct that is discovered by a set entry point could be useful). In any case, signatures are not really on the critical path anyway. So it is not that important. I did leave a comment asking for a broad sketch of how it works. 

Otherwise, I left low level comments on the implementation of ExecutionSession.

File: src/Runtime/ExecutionSession.cpp (hunk 11 above)
Comment (tungld) [reply src/Runtime/ExecutionSession.cpp]: Done for this and other functions too. Thanks!

File: src/Runtime/ExecutionSession.hpp (hunk 12 above)
Comment (tungld) [reply src/Runtime/ExecutionSession.hpp]: Yes, I changed the order.

File: src/Runtime/ExecutionSession.cpp (hunk 13 above)
Comment (tungld) [reply src/Runtime/ExecutionSession.cpp:130]: Done.

File: test/backend-cpp/ModelBuilder.cpp (hunk 15 above)
Comment (tungld) [reply test/backend-cpp/ModelBuilder.cpp]: Yes, it's redundant. Done.

File: include/onnx-mlir/Runtime/OMSignature.h (hunk 14 above)
Comment (tungld) [reply include/onnx-mlir/Runtime/OMSignature.h:27]: LLVM code is equivalent to the following pseudo code:
```C
entry_point_0 = "run_add";
entry_point_0_in_sig = "[....]"; // JSON string.
entry_point_1 = "run_sub";
entry_point_1_in_sig = "[....]"; // JSON string.
const char* omInputSignature(entry_point) {
    if (entry_point == entry_point_1)
      return entry_point_0_in_sig;
    if (entry_point == entry_point_1)
      return entry_point_1_in_sig;
    return undefined;
}
```

We do not see if-then-else but branches at LLVM level.

File: test/numerical/numerical.def (hunk 16 above)
Comment (tungld) [reply test/numerical/numerical.def:1]: Done for other .def files. Thanks!

File: src/Runtime/ExecutionSession.hpp (hunk 17 above)
Comment (tungld) [reply src/Runtime/ExecutionSession.hpp]: Done. Now we use a single constructor.

File: include/onnx-mlir/Runtime/OMSignature.h (hunk 14 above)
Comment (AlexandreEichenberger) [reply include/onnx-mlir/Runtime/OMSignature.h:27]: Cool, and the compare `entry_point == entry_point_0` simply compare the addresses of the strings?

File: test/backend-cpp/ModelBuilder.cpp (hunk 15 above)
Comment (AlexandreEichenberger) [reply test/backend-cpp/ModelBuilder.cpp]: Its fine if you wanted to test the functionality, but if you do this, just add a comment as otherwise folks may think that it is still needed.

File: include/onnx-mlir/Runtime/OMSignature.h (hunk 14 above)
Comment (tungld) [reply include/onnx-mlir/Runtime/OMSignature.h:27]: Yes, the address of the first char.

File: src/Runtime/ExecutionSession.hpp [hunk 18]
```diff
@@ -24,15 +24,21 @@
...
 public:
-  ExecutionSession(std::string sharedLibPath);
-  ExecutionSession(std::string sharedLibPath, std::string entryPointName);
+  ExecutionSession(std::string sharedLibPath, bool defaultEntryPoint = true);
+
+  // Get a NULL-terminated array of entry point names.
+  // For example {"run_addition, "run_substraction", NULL}
+  const std::string *queryEntryPoints() const;
+
+  // Set entry point for this session.
```
Comment (AlexandreEichenberger) [inline src/Runtime/ExecutionSession.hpp:40]: Please just add in the comments that this call is necessary when the defaultEntryPoint is not used, or if it was used but the user wants now to use a different entry point.  

I have a question related to the string passed to the setEntryPoint. I know that for the "run" part, a dlsym is used, so the string chars are used to locate the symbol. But for the signatures, is it the address of the string that needs to be the same as recorded, or is it the content of the string?

In other words, are either of these calls working?

``` C++
setEntryPoint(queryEntryPoints()[0]); // use string directly
setEntryPoint(string(queryEntryPoints()[0])); // use a copy of the string
```

**Review by AlexandreEichenberger — COMMENTED** [review]
Thanks for the code changes. Still have a question pending that has an impact on how the interface works. Once that is resolved, it will be all good for me.

File: src/Runtime/ExecutionSession.hpp (hunk 18 above)
Comment (tungld) [reply src/Runtime/ExecutionSession.hpp:40]: > setEntryPoint(queryEntryPoints()[0]); // use string directly
> setEntryPoint(string(queryEntryPoints()[0])); // use a copy of the string
> is it the content of the string?

Yes, it is the string content. so both work.

Comment (tungld) [reply src/Runtime/ExecutionSession.hpp:40]: > Please just add in the comments that this call is necessary when the defaultEntryPoint is not used, or if it was used but the user wants now to use a different entry point.

Done this. Thanks!

File: include/onnx-mlir/Runtime/OMSignature.h (hunk 14 above)
Comment (AlexandreEichenberger) [reply include/onnx-mlir/Runtime/OMSignature.h:27]: This seems at odd with your comment that the comparison is done on the string content. Can you run a test to make sure that works? Thanks

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for all the updates and answers to questions.
